README.md
# `[DEMO]` cloudflare-durable-object-sessions

Demo of Cloudflare Workers and Durable Objects as an approach to user and session management.

This was just me, trying to understand how Durable Objects work, and how I might use them for authentication and session management.

Docs of interest:

- Learning the basics: https://developers.cloudflare.com/workers/learning/using-durable-objects
- The runtime APIs: https://developers.cloudflare.com/workers/runtime-apis/durable-objects
- Using the CF API: https://api.cloudflare.com/#durable-objects-namespace-list-objects

## Deploy notes

You can run this yourself, if you want. Make sure to change the `routes` in the `wrangler.toml` file,
and everything else should work.

At the end, to delete the whole stack you'll need to run a Wrangler migration to delete the
Durable Object class (see the `wrangler.toml` file for notes), and then you can go into
the Worker page (in the Cloudflare dashboard), into "Manage Services", and there
delete all the stuff created here.

## License

Published and released under the [Very Open License](http://veryopenlicense.com).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Change categories are:

* `Added` for new features.
* `Changed` for changes in existing functionality.
* `Deprecated` for once-stable features removed in upcoming releases.
* `Removed` for deprecated features removed in this release.
* `Fixed` for any bug fixes.
* `Security` to invite users to upgrade in case of vulnerabilities.

## [Unreleased]
### Added
### Changed
### Deprecated
### Fixed
### Removed
### Security

## [0.0.0](https://github.com/saibotsivad/cloudflare-durable-object-sessions/tree/v0.0.0) - 2022-01-17
### Added
- Created the base project from [saibotsivad/init](https://github.com/saibotsivad/init).

[Unreleased]: https://github.com/saibotsivad/cloudflare-durable-object-sessions/compare/v0.0.0...HEAD
[0.0.1]: https://github.com/saibotsivad/cloudflare-durable-object-sessions/compare/v0.0.0...v0.0.1


src/index.test.js
import { test } from 'uvu'
import * as assert from 'uvu/assert'
import TODO from './index.js'

test('TODO', () => {
	assert.type(TODO, 'function')
})

test.run()


src/counter.js
export class Counter {
	constructor(state, env) {
		this.state = state
		// `blockConcurrencyWhile()` ensures no requests are delivered until
		// initialization completes.
		this.state.blockConcurrencyWhile(async () => {
			let stored = await this.state.storage.get("value")
			this.value = stored || 0
		})
	}

	// Handle HTTP requests from clients.
	async fetch(request) {
		// Apply requested action.
		let url = new URL(request.url)
		let currentValue = this.value
		switch (url.pathname) {
		case "/increment":
			currentValue = ++this.value
			await this.state.storage.put("value", this.value)
			break
		case "/decrement":
			currentValue = --this.value
			await this.state.storage.put("value", this.value)
			break
		case "/":
			// Just serve the current value. No storage calls needed!
			break
		default:
			return new Response("Not found", { status: 404 })
		}

		// Return `currentValue`. Note that `this.value` may have been
		// incremented or decremented by a concurrent request when we
		// yielded the event loop to `await` the `storage.put` above!
		// That's why we stored the counter value created by this
		// request in `currentValue` before we used `await`.
		return new Response(currentValue)
	}
}


src/objects/user.js
const commands = {
	profile: {
		get: async ({ storage }) => storage.get('profile'),
		put: async ({ storage }, { username, password }) => storage.put('profile', { username, password }),
		del: async ({ storage }) => storage.delete('profile'),
	},
	session: {
		get: async ({ storage }, { id }) => storage.get(`session|${id}`),
		put: async ({ storage }, { id, secret, expired }) => storage.put(`session|${id}`, {
			id,
			secret,
			expired,
		}).then(() => ({ id, secret, expired })),
		del: async ({ storage }, { id }) => storage.delete(`session|${id}`),
	},
	sessions: {
		list: async ({ storage }, opts) => {
			const params = { prefix: 'session|' }
			if (opts?.limit) params.limit = opts.limit
			if (opts?.start) params.start = opts.start
			if (opts?.reverse) params.reverse = opts.reverse
			return storage
				.list(params)
				.then(map => Object.fromEntries(map))
		},
	},
}

export class User {
	constructor(state, env) {
		this.state = state
	}

	async fetch(request) {
		const body = request.method === 'POST' && request.headers.get('content-type')?.includes('application/json') && await request.json()
		/*
		The body here is expected to look like:

			[
				"profile",
				"put",
				{ "username": "saibotsivad", ...etc }
			]
		 */
		let command = commands[body?.[0]]?.[body?.[1]]
		if (command) {
			const result = await command(this.state, body[2])
			return new Response(JSON.stringify({ data: result }), {
				status: 200,
				headers: { 'Content-Type': 'application/json' },
			})
		}
		return new Response(JSON.stringify({ ok: false }), {
			status: 404,
			headers: { 'Content-Type': 'application/json' },
		})
	}

	// another alternate approach would be maybe something like
	getProfile() {
		return this.state.storage.get('profile')
	}
	/*
	Then you could do something like

	async fetch(request) {
		const url = new URL(request.url)
		const [ , methodName ] = url.pathname.split('/)
		this[methodName](await request.json())
	}
	 */
}


src/index.js
import App from './routes/App.svelte'
import Login from './routes/Login.svelte'
import LoggedOut from './routes/LoggedOut.svelte'
import NewAccount from './routes/NewAccount.svelte'
import Profile from './routes/Profile.svelte'
import * as Cookie from 'worktop/cookie'
import * as Base64 from 'worktop/base64'
import { ulid, uid } from 'worktop/utils'
import Trouter from 'trouter'

const html = ({ head, html, css }) => `<!DOCTYPE html>
<head lang="en-US">
	<meta name="viewport" content="width=device-width" />
	<meta charset="UTF-8" />
	<style>
		body { background-color: #e2e2e2; }
		${css.code}
	</style>
	<title>CF DO Sessions Demo</title>
	${head}
</head>
<body>
${html}
</body>`

const htmlResponse = (status, component, headers = {}) => new Response(
	html(component),
	{
		status,
		headers: {
			'Content-Type': 'text/html;charset=UTF-8',
			...headers,
		},
	},
)

const readForm = async request => {
	const formData = await request.formData()
	const body = {}
	for (const entry of formData.entries()) {
		body[entry[0]] = entry[1]
	}
	return body
}

const parseCookie = request => {
	try {
		return JSON.parse(
			Base64.decode(
				Cookie.parse(request.headers.get('cookie') || '').SESSION,
			),
		)
	} catch (ignore) {
		//
	}
	return null
}

const stringifyCookie = params => Cookie.stringify(
	'SESSION',
	Base64.encode(
		JSON.stringify(params),
	),
	// TODO httpOnly etc
)

const makeDurableFetch = (url, durableClass) => async (id, command, action, params) => {
	const durable = durableClass.get(durableClass.idFromName(id))
	const response = await durable.fetch(url, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify([ command, action, params ]),
	})
	try {
		return (await response.json()).data
	} catch (error) {
		console.log('error parsing json response', { id, command, action, params })
		throw error
	}
}

const router = new Trouter()

router
	.get('/', async () => htmlResponse(200, App.render()))
	.get('/login', async ({ url, user, session }) => {
		if (user && session) return Response.redirect(`${url.origin}/profile`)
		return htmlResponse(200, Login.render())
	})
	.post('/login', async ({ url, form, durableFetch }) => {
		if (!form.username || !form.password) return Response.redirect(`${url.origin}/login?message=${encodeURIComponent('No username/password specified.')}`)
		const loginUser = await durableFetch(form.username, 'profile', 'get', undefined)
		// TODO use hash instead
		if (loginUser && loginUser.password === form.password) {
			const session = await durableFetch(form.username, 'session', 'put', {
				id: ulid(),
				secret: uid(32),
			})
			return session?.id
				? new Response('Redirecting...', {
					status: 303,
					headers: {
						'Set-Cookie': stringifyCookie({
							id: session.id,
							secret: session.secret,
							username: form.username,
						}),
						Location: `${url.origin}/profile`,
					},
				})
				: Response.redirect(`${url.origin}/login?message=${encodeURIComponent('Error while saving session.')}`)
		}
		return Response.redirect(`${url.origin}/login?message=${encodeURIComponent('Bad username or password. ')}`)
	})
	.get('/profile', async ({ url, durableFetch, user, session }) => {
		return user
			? htmlResponse(200, Profile.render({
				user,
				session,
				sessions: await durableFetch(user.username, 'sessions', 'list'),
			}))
			: Response.redirect(`${url.origin}/login`)
	})
	.post('/profile', async ({ url, durableFetch, user, form }) => {
		if (!user) return Response.redirect(`${url.origin}/login`)
		if (!form?.password) return Response.redirect(`${url.origin}/profile?message=${encodeURIComponent('No password provided.')}`)
		user.password = form.password // TODO hash+salt before saving
		await durableFetch(user.username, 'profile', 'put', user)
		return Response.redirect(`${url.origin}/profile?message=${encodeURIComponent('Password has been changed.')}`)
	})
	.get('/new_account', async ({ url, user }) => {
		if (user) return Response.redirect(`${url.origin}/profile`)
		return htmlResponse(200, NewAccount.render())
	})
	.post('/new_account', async ({ form, url, durableFetch }) => {
		if (!form.username || !form.password) return Response.redirect(`${url.origin}/new_account?message=${encodeURIComponent('No username/password specified.')}`)
		const existingUser = await durableFetch(form.username, 'profile', 'get')
		if (existingUser) return Response.redirect(`${url.origin}/new_account?message=${encodeURIComponent('Username is already taken.')}`)
		await durableFetch(form.username, 'profile', 'put', {
			username: form.username,
			password: form.password, // TODO hash+salt before saving
		})
		return Response.redirect(`${url.origin}/login?message=${encodeURIComponent('User created, go ahead and log in.')}`)
	})
	.get('/logout', async ({ user, session, durableFetch }) => {
		if (user && session) await durableFetch(user.username, 'session', 'put', {
			id: session.id,
			expired: Date.now(),
		})
		return htmlResponse(200, LoggedOut.render(), { 'Set-Cookie': stringifyCookie({}) })
	})
	.get('/session/remove/:sessionId', async ({ url, user, durableFetch, sessionId }) => {
		if (user) await durableFetch(user.username, 'session', 'del', { id: sessionId })
		return Response.redirect(`${url.origin}/profile?message=${encodeURIComponent('Deleted session: ' + sessionId)}`)
	})

async function handleRequest(request, env) {
	const url = new URL(request.url)
	const cookie = parseCookie(request)

	// The `env.USER` is how you bind an exported class name, in this case
	// the `export { User }` to a Worker instantiation environment. Look at
	// the `wrangler.toml` file for more details to see how `User` maps to `USER`.
	const durableFetch = makeDurableFetch(url.origin + '/DO', env.USER)

	/*
	This `durableFetch` is wrapping up this boilerplate:

	// This is how you instantiate a Durable Object:
	const username = '' // from a form, session cookie, auth token, etc.
	const durableId = env.USER.idFromName(username)
	const durableInstance = env.USER.get(durableId)

	// To interact with it, you use `instance.fetch` which behaves the
	// same as the `globalThis.fetch`. The `url` doesn't actually matter,
	// but if you are using a path router in the Durable Object (I'm not
	// in this demo) you'll of course need to set it here.
	//
	// Note that the URL used here will show up in the Worker logs, so
	// make sure to be careful with privacy concerns.
	const response = durableInstance.fetch(url.origin + '/', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify([ command, action, params ]),
	})

	// The `response` is a normal `fetch` response, so you can just
	// grab the JSON, if that's what came back:
	const data = await response.json()

	 */

	let user
	let session
	if (cookie && cookie.id && cookie.username && cookie.secret) {
		session = await durableFetch(cookie.username, 'session', 'get', { id: cookie.id })
		if (session && session.secret && session.secret === cookie.secret) {
			user = await durableFetch(cookie.username, 'profile', 'get')
		}
	}

	const form = request.method === 'POST'
		&& request.headers.get('content-type')?.includes('form')
		&& await readForm(request)

	const route = router.find(request.method, url.pathname)
	if (route.handlers.length) {
		return route.handlers[0]({
			url,
			user,
			session,
			form,
			durableFetch,
			...route.params,
		})
	} else {
		return new Response('Not found', { status: 404 })
	}
}

export default {
	async fetch(request, env) {
		try {
			return await handleRequest(request, env)
		} catch (e) {
			return new Response(e.message)
		}
	},
}

export { User } from './objects/user.js'
